File: src/test-misc/csharp-spec2/MainClass.cs
using Csharp.Spec2smTests;
using System;

namespace StateSmithTest.spec._2.CSharp;

public class MainClass
{
    public static void Trace(string str)
    {
        Console.WriteLine(str);
        Console.Out.Flush();      // might be useful in case code fails in someway
    }

    public static bool trace_guard(string str, bool guard)
    {
        Console.Write(str + " ");

        if (guard)
        {
            Console.Write("Behavior running.");
        }
        else
        {
            Console.Write("Behavior skipped.");
        }

        Console.Write("\n");

        return guard;
    }

    public static void print_divider()
    {
        Console.Write("===================================================\n");
    }

    public static void print_start()
    {
        Console.Write("Start Statemachine\n");
        print_divider();
    }

    public static void print_dispatch_event_name(string event_name)
    {
        Console.Write($"Dispatch event {event_name}\n");
        print_divider();
    }

    public static void ActualMain(string[] args)
    {
        Spec2Sm sm = new();

        print_start();
        sm.Start();
        Console.WriteLine();

        for (int i = 0; i < args.Length; i++) // start at 1 to skip program name
        {
            string arg = args[i];
            
            if (!Enum.TryParse(arg.ToUpper(), out Spec2Sm.EventId eventId))
            {
                throw new ArgumentException($"bad i:{i} arg: `{arg}`");
            }

            print_dispatch_event_name(arg);

            sm.DispatchEvent(eventId);
            Console.WriteLine();
        }
    }

    public static void Main(string[] args)
    {
        ActualMain(args);
    }
}

File: src/test-misc/swift-spec2/Sources/MainClass.swift
@MainActor
public class MainClass {
    static let eventMapping: [String: Spec2Sm.EventId] = [
        "DO": .DO,
        "EV1": .EV1,
        "EV10": .EV10,
        "EV2": .EV2,
        "EV3": .EV3,
        "EV4": .EV4,
        "EV5": .EV5,
        "EV6": .EV6,
        "EV7": .EV7,
        "EV8": .EV8,
        "EV9": .EV9,
        "EVBACK": .EVBACK,
        "EVCLOSE": .EVCLOSE,
        "EVOPEN": .EVOPEN,
        "EVSTEP": .EVSTEP,
    ]

    public static func trace(_ str: String) {
        print(str)
    }

    public static func traceGuard(_ str: String, _ `guard`: Bool) -> Bool {
        print(str + " ", terminator:"")

        if `guard` {
            print("Behavior running.", terminator:"")
        } else {
            print("Behavior skipped.", terminator:"")
        }

        print("")

        return `guard`
    }

    public static func printDivider() {
        print("===================================================")
    }

    public static func printStart() {
        print("Start Statemachine")
        printDivider()
    }

    public static func printDispatchEventName(_ eventName: String) {
        print("Dispatch event " + eventName)
        printDivider()
    }

    public static func actualMain(_ args: [String]) {
        let sm = Spec2Sm()

        printStart()
        sm.start()
        print("")

        for arg in args {
            guard let eventId = eventMapping[arg] else {
                print("bad arg: `" + arg + "`")
            }
            printDispatchEventName(arg)
            sm.dispatchEvent(eventId)
            print("")
        }
    }
}

@main
struct SpecSwift {
    static func main() {
        MainClass.actualMain(CommandLine.arguments)
    }
}